File: smalivm/src/main/java/org/cf/smalivm/opcode/AGetOpFactory.java
package org.cf.smalivm.opcode;

import gnu.trove.map.TIntObjectMap;

import org.cf.smalivm.VirtualMachine;
import org.cf.util.Utils;
import org.jf.dexlib2.builder.MethodLocation;
import org.jf.dexlib2.iface.instruction.formats.Instruction23x;

public class AGetOpFactory implements OpFactory {

    @Override
    public Op create(MethodLocation location, TIntObjectMap<MethodLocation> addressToLocation, VirtualMachine vm) {
        MethodLocation child = Utils.getNextLocation(location, addressToLocation);
        Instruction23x instr = (Instruction23x) location.getInstruction();
        int valueRegister = instr.getRegisterA();
        int arrayRegister = instr.getRegisterB();
        int indexRegister = instr.getRegisterC();

        return new AGetOp(location, child, valueRegister, arrayRegister, indexRegister, vm.getExceptionFactory());
    }

}

File: smalivm/src/main/java/org/cf/smalivm/opcode/AGetOpFactory.kt
package org.cf.smalivm.opcode

import org.cf.smalivm.ExceptionFactory
import org.cf.smalivm.type.ClassManager
import org.cf.util.Utils
import org.jf.dexlib2.builder.MethodLocation
import org.jf.dexlib2.iface.instruction.formats.Instruction23x

class AGetOpFactory : OpFactory {
    override fun create(location: MethodLocation, addressToLocation: Map<Int, MethodLocation>, classManager: ClassManager, exceptionFactory: ExceptionFactory): Op {
        val child = Utils.getNextLocation(location, addressToLocation)
        val instr = location.instruction as Instruction23x
        val valueRegister = instr.registerA
        val arrayRegister = instr.registerB
        val indexRegister = instr.registerC
        return AGetOp(location, child, valueRegister, arrayRegister, indexRegister, vm.exceptionFactory)
    }
}
